pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategicpatch
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+)
+
+// An alternate implementation of JSON Merge Patch
+// (https://tools.ietf.org/html/rfc7386) which supports the ability to annotate
+// certain fields with metadata that indicates whether the elements of JSON
+// lists should be merged or replaced.
+//
+// For more information, see the PATCH section of docs/api-conventions.md.
+func StrategicMergePatchData(original, patch []byte, dataStruct interface{}) ([]byte, error) {
+	var o map[string]interface{}
+	err := json.Unmarshal(original, &o)
+	if err != nil {
+		return nil, err
+	}
+
+	var p map[string]interface{}
+	err = json.Unmarshal(patch, &p)
+	if err != nil {
+		return nil, err
+	}
+
+	t := reflect.TypeOf(dataStruct)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	if t.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("strategic merge patch needs a struct, %s received instead", t.Kind().String())
+	}
+
+	result, err := mergeMap(o, p, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(result)
+}
+
+const specialKey = "__patch"
```

Comment:
My interpretation of the spec is that '$` and `.` are legal map key characters.

What do you think about `$patch` (similar to `$schema` in json schema)?

FWIW, I found others also using `$` for operators:
http://docs.mongodb.org/manual/reference/operator/update/positional/